Mathlib/Data/Real/ConjExponents.lean

```diff
@@ -0,0 +1,234 @@
+/-
+Copyright (c) 2020 Sébastien Gouëzel. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sébastien Gouëzel, Yury Kudryashov
+-/
+import Mathlib.Data.ENNReal.Real
+
+#align_import data.real.conjugate_exponents from "leanprover-community/mathlib"@"2196ab363eb097c008d4497125e0dde23fb36db2"
+
+/-!
+# Real conjugate exponents
+
+This file defines conjugate exponents in `ℝ` and `ℝ≥0`. Real numbers `p` and `q` are *conjugate* if
+they are both greater than `1` and satisfy `p⁻¹ + q⁻¹ = 1`. This property shows up often in
+analysis, especially when dealing with `L^p` spaces.
+
+## Main declarations
+
+* `Real.IsConjExponent`: Predicate for two real numbers to be conjugate.
+* `Real.conjExponent`: Conjugate exponent of a real number.
+* `NNReal.IsConjExponent`: Predicate for two nonnegative real numbers to be conjugate.
+* `NNReal.conjExponent`: Conjugate exponent of a nonnegative real number.
+
+## TODO
+
+* Eradicate the `1 / p` spelling in lemmas.
+* Do we want an `ℝ≥0∞` version?
+-/
+
+noncomputable section
+
+open scoped ENNReal
+
+namespace Real
+
+/-- Two real exponents `p, q` are conjugate if they are `> 1` and satisfy the equality
+`1/p + 1/q = 1`. This condition shows up in many theorems in analysis, notably related to `L^p`
+norms. -/
+@[mk_iff]
+structure IsConjExponent (p q : ℝ) : Prop where
```

Comment:
Is there a typeclass that abstracts this? I guess it would be something like `OrderedSemifield`... which we probably don't have.

Reply:
We do have `LinearOrderedSemifield`. Do you think that's worth it given that there are exactly two type which would fit the bill (namely `ℝ` and `ℝ≥0`. `ℝ≥0∞` is not a linear ordered semifield so would need to be handled separately)?

Reply:
If it roughly halves the length of this file, then yes I would attempt to use it, I think.

If we can come up with a typeclass that also humours `ENNReal`, that would be even better.

Reply:
I can't, actually. There's too much about `tsub` that's not generalised (generalisable?) so that it applies to both `Real` and `NNReal`.

Reply:
Ok, good point.